contracts/javascore/aggregator/src/main/java/relay/aggregator/RelayAggregator.java

```diff
@@ -0,0 +1,305 @@
+            }
+        }
+    }
+
+    @External
+    public void removeRelayers(Address[] relayersToBeRemoved) {
+        adminOnly();
+
+        Context.require(relayersToBeRemoved != null && relayersToBeRemoved.length != 0,
+                "relayers to be removed cannot be empty");
+
+        HashMap<Address, Integer> existingRelayers = new HashMap<Address, Integer>();
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            existingRelayers.put(relayer, i);
+        }
+
+        for (Address relayerToBeRemoved : relayersToBeRemoved) {
+            if (existingRelayers.containsKey(relayerToBeRemoved)) {
+                Address top = relayers.pop();
+                if (!top.equals(relayerToBeRemoved)) {
+                    Integer pos = existingRelayers.get(relayerToBeRemoved);
+                    relayers.set(pos, top);
+                }
+                existingRelayers.remove(relayerToBeRemoved);
+            }
+        }
+    }
+
+    @External
+    public void registerPacket(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            BigInteger srcHeight,
+            String dstNetwork,
+            byte[] data) {
+
+        adminOnly();
+
+        Packet pkt = new Packet(srcNetwork, contractAddress, srcSn, srcHeight, dstNetwork, data);
+        String id = pkt.getId();
+
+        Context.require(packets.get(id) == null, "Packet already exists");
+
+        packets.set(id, pkt);
+
+        PacketRegistered(
+                pkt.getSrcNetwork(),
+                pkt.getContractAddress(),
+                pkt.getSrcSn(),
+                pkt.getSrcHeight(),
+                pkt.getDstNetwork(),
+                pkt.getData());
+    }
+
+    @External
+    public void acknowledgePacket(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            byte[] signature) {
+
+        relayersOnly();
+
+        String pktID = Packet.createId(srcNetwork, contractAddress, srcSn);
+        Packet pkt = packets.get(pktID);
+        Context.require(pkt != null, "Packet not registered");
+
+        byte[] existingSign = signatures.at(pktID).get(Context.getCaller());
+        Context.require(existingSign == null, "Signature already exists");
+
+        setSignature(pktID, Context.getCaller(), signature);
+
+        if (signatureThresholdReached(pktID)) {
+            byte[][] sigs = getSignatures(srcNetwork, contractAddress, srcSn);
+            byte[] encodedSigs = serializeSignatures(sigs);
+            PacketAcknowledged(
+                    pkt.getSrcNetwork(),
+                    pkt.getContractAddress(),
+                    pkt.getSrcSn(),
+                    pkt.getSrcHeight(),
+                    pkt.getDstNetwork(),
+                    pkt.getData(),
+                    encodedSigs);
+            removePacket(pktID);
+        }
+    }
+
```

Comment:
instead of register and acknowledge we can merge it to single method say relay packet. The leader will submit the packet with signatures , if the id is new then we emit PacketRegistered if the id already exists we can emit acknowledge event. This will simplify logic.

Reply:
I think this would rather complicate the logic and need to handle additional cases.